src/libarena/lib.rs

```diff
@@ -106,7 +106,7 @@ impl<T> TypedArenaChunk<T> {
         unsafe {
             if mem::size_of::<T>() == 0 {
                 // A pointer as large as possible for zero-sized elements.
-                !0 as *mut T
+                (!0usize) as *mut T
```

Comment:
@bluss It appears that `!0 as *mut T` has a magical power to fix the type of the integer literal to `usize`. It looks more and more suspicious when we see that [LLVM `inttoptr`](http://llvm.org/docs/LangRef.html#constant-expressions) is defined as zero extention.

- https://play.rust-lang.org/?gist=74d0b36e704a3f7dc4c686cd3ab49827&version=nightly&backtrace=0